_posts/2023-07-14-types-and-programming-languages.markdown

```diff
@@ -0,0 +1,610 @@
+
+That concludes the _progress_ part of soundness checking: whenever a term type-checks, there is always an applicable small step evaluation rule unless it's a final value.
+
+#### The preservation check
+
+Soundness has another aspect, preservation, as stated in the intro. It says that, when evaluating a well-typed term, the evaluator will not get stuck and the result will have the same type as the original term.
+Dafny can also prove it for our language, out of the box. Well done, that means our language and evaluator make sense together!
+
+{% highlight javascript %}
+lemma OneStepEvaluateWellTyped(e: Term)
+  requires WellTyped(e) && !IsFinalValue(e)
+  ensures GetType(OneStepEvaluate(e)) == GetType(e)
+{
+  // Amazing: Dafny can prove this soundness property automatically
+}
+{% endhighlight %}
+
+#### Conclusion
+
+All the code above powers this page, which is why I can guarantee you that you won't be able to find a term that the type checker accepts and that won't result in a final value. Of course, in a real programming language term, you might add some infinite loops, but the soundness property above is not about termination, it's about constant progress, which you also want in embedded systems to ensure they never need reboot.
+
+Now that you know what a type checker is and how to implement one in Dafny, perhaps you will feel much better prepared to model and experiment on your new programming language, like recently the [Cedar team did](https://github.com/cedar-policy/cedar-spec)?
+
+This is the end of the blog post. I hope you enjoyed it!
+
+---
+
+## Bonus: more advanced modeling
+If you are looking for some advanced concepts, feel free to continue reading! Beware, math ahead!
+
+Sometimes, modeling the evaluator and the type-checker as functions is not enough. One wants to model them as relations, and determine some properties about these relations, such as the order of evaluation being irrelevant for the final result.
+
+In the rest of this blog post, largely inspired by the book "Types and Programming Languages", Chapter 8, written by Benjamin Pierce, I will illustrate one element of the proof: the one that inductive and constructive versions of the set of Terms are equivalent. Having equivalence enables obtaining other results out of the scope of this blog post, including that the order of evaluation does not matter.
+
+With the help of this trick, it becomes possible to prove similar equivalences for different inductive and constructive definitions of:
+- The set of `(Expr, Expr)` of small-step evaluations
+- The set of `(Expr, Type)` of type checking
+
+but I leave these as an exercise for the interested reader.
+
+In Types and Programming Languages, chapter 3.2, we discover that there are two other mathematical definitions of the "set of all terms".
+The first one in definition 3.2.1 states that the set of _terms_ is the smallest set 𝒯 such that:
```

Comment:
In the smallest set definition, you use two different cursive Ts. (I did notice when copying the T in text it appears as 𝒯, and different in math)